bentoml/_internal/server/grpc/interceptors/prometheus.py

```diff
@@ -0,0 +1,116 @@
+from __future__ import annotations
+
+import typing as t
+import logging
+import functools
+import contextvars
+from timeit import default_timer
+from typing import TYPE_CHECKING
+
+from grpc import aio
+from simple_di import inject
+from simple_di import Provide
+
+from ....utils.grpc import wrap_rpc_handler
+from ....configuration.containers import BentoMLContainer
+
+if TYPE_CHECKING:
+    from ..types import Request
+    from ..types import Response
+    from ..types import RpcMethodHandler
+    from ..types import AsyncHandlerMethod
+    from ..types import HandlerCallDetails
+    from ..types import BentoServicerContext
+    from ....service import Service
+    from ...metrics.prometheus import PrometheusClient
+
+logger = logging.getLogger(__name__)
+START_TIME_VAR: contextvars.ContextVar[float] = contextvars.ContextVar("START_TIME_VAR")
+
+
+class PrometheusServerInterceptor(aio.ServerInterceptor):
+    """
+    An async interceptor for prometheus metrics
+    """
+
+    @inject
+    def __init__(
+        self,
+        bento_service: Service,
+        metrics_client: PrometheusClient = Provide[BentoMLContainer.metrics_client],
+    ):
+        self.metrics_client = metrics_client
+
+        self.service_name = bento_service.name
+        # a valid tag name may includes invalid characters, so we need to escape them
+        # ref: https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels
+        self.service_name = self.service_name.replace("-", ":").replace(".", "::")
+
+        self.service_version = (
+            bento_service.tag.version if bento_service.tag is not None else ""
+        )
+
+        self.metrics_request_duration = metrics_client.Histogram(
+            name=f"{self.service_name}_request_duration_seconds",
+            documentation=f"{self.service_name} API GRPC request duration in seconds",
+            labelnames=["api_name", "service_version", "grpc_response_code"],
+        )
+        self.metrics_request_total = metrics_client.Counter(
+            name=f"{self.service_name}_request_total",
+            documentation="Total number of GRPC requests",
+            labelnames=["api_name", "service_version", "grpc_response_code"],
```

Comment:
Instead of `grpc_response_code`, can we do `http_response_code` here? It is easier understand.